File: clients/android/GatewayClient.kt
// clients/android/GatewayClient.kt
// In app/build.gradle.kts:
//   implementation("com.squareup.okhttp3:okhttp:4.12.0")
//   implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.8.1")

package com.example.gateway

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.security.KeyFactory
import java.security.PublicKey
import java.security.SecureRandom
import java.security.spec.MGF1ParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PSource
import javax.crypto.spec.SecretKeySpec

data class GatewayResult(
    val statusCode: Int,
    val headers: Map<String, String>,
    val data: Any?
)

/** Gateway error (400/429/502/504). `code`: UNKNOWN_KEY_ID, ... */
class GatewayException(val status: Int, val code: String?, message: String) : Exception(message)

class GatewayClient(
    private val gatewayUrl: String,
    serverPublicKeyPem: String,
    private val keyId: String? = null,
    private val http: OkHttpClient = OkHttpClient(),
    private val requireEncryptedResponse: Boolean = true,
) {
    private val serverKey: PublicKey = parsePublicKeyPem(serverPublicKeyPem)
    private val random = SecureRandom()

    /** Sends a request through the encryption gateway. */
    suspend fun send(
        method: String,
        accessPoint: String,
        headers: Map<String, String> = emptyMap(),
        body: Any? = null, // JSONObject, JSONArray, or String
        parameter: JSONObject? = null,
    ): GatewayResult = withContext(Dispatchers.IO) {
        val payload = JSONObject()
            .put("Method", method)
            .put("AccessPoint", accessPoint)
            .put("Header", JSONArray(headers.map { (k, v) -> JSONObject().put("Key", k).put("Value", v) }))
        if (body != null) payload.put("Body", body)
        if (parameter != null) payload.put("Parameter", parameter)

        // 1. Fresh AES key + IV for this request; 2. AES-256-GCM payload
        val aesKey = SecretKeySpec(randomBytes(32), "AES")
        val iv = randomBytes(12)
        val sealed = Cipher.getInstance("AES/GCM/NoPadding").run {
            init(Cipher.ENCRYPT_MODE, aesKey, GCMParameterSpec(128, iv))
            doFinal(payload.toString().toByteArray(Charsets.UTF_8))
        }

        // 3. RSA-OAEP-SHA256 AES key (with MGF1-SHA256)
        val encryptedKey = Cipher.getInstance("RSA/ECB/OAEPPadding").run {
            init(Cipher.ENCRYPT_MODE, serverKey, OAEP_SHA256)
            doFinal(aesKey.encoded)
        }

        val requestObj = JSONObject()
            .put("EncryptedKey", b64(encryptedKey))
            .put("IV", b64(iv))
            .put("Payload", b64(sealed))
        if (keyId != null) requestObj.put("KeyId", keyId)

        val request = Request.Builder()
            .url(gatewayUrl)
            .post(requestObj.toString().toRequestBody("application/json".toMediaType()))
            .build()

        http.newCall(request).execute().use { res ->
            val text = res.body?.string().orEmpty()
            val json = runCatching { JSONObject(text) }.getOrNull()
            if (res.code != 200) {
                throw GatewayException(
                    res.code,
                    json?.optString("code")?.ifEmpty { null },
                    json?.optString("message") ?: text
                )
            }

            val envelope = json ?: throw GatewayException(200, "INVALID_RESPONSE", "Invalid gateway response")
            val marker = envelope.opt("Encrypted")
            val result = if (marker == false) {
                if (requireEncryptedResponse) {
                    throw GatewayException(200, "ENCRYPTION_REQUIRED", "Encrypted response required; check ENCRYPT_RESPONSE on the server")
                }
                if (envelope.has("IV") || envelope.has("Payload")) {
                    throw GatewayException(200, "INVALID_RESPONSE", "Invalid gateway response")
                }
                envelope
            } else {
                if (marker != true && envelope.has("Encrypted")) {
                    throw GatewayException(200, "INVALID_RESPONSE", "Invalid gateway response")
                }
                // Legacy encrypted envelopes without a marker are also supported.
                val plain = Cipher.getInstance("AES/GCM/NoPadding").run {
                    init(Cipher.DECRYPT_MODE, aesKey, GCMParameterSpec(128, unb64(envelope.getString("IV"))))
                    doFinal(unb64(envelope.getString("Payload")))
                }
                JSONObject(String(plain, Charsets.UTF_8))
            }
            val statusCode = result.opt("StatusCode")
            if (statusCode !is Int || statusCode !in 100..599 ||
                result.optJSONObject("Headers") == null || !result.has("Data")) {
                throw GatewayException(200, "INVALID_RESPONSE", "Invalid gateway response")
            }

            val responseHeaders = mutableMapOf<String, String>()
            val rawHeaders = result.optJSONObject("Headers")
            if (rawHeaders != null) {
                for (k in rawHeaders.keys()) {
                    responseHeaders[k] = rawHeaders.getString(k)
                }
            }

            GatewayResult(
                statusCode = statusCode,
                headers = responseHeaders,
                data = result.opt("Data")
            )
        }
    }

    private fun randomBytes(n: Int) = ByteArray(n).also { random.nextBytes(it) }

    private companion object {
        val OAEP_SHA256 = OAEPParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec.SHA256, PSource.PSpecified.DEFAULT)

        fun b64(bytes: ByteArray): String = Base64.getEncoder().encodeToString(bytes)
        fun unb64(text: String): ByteArray = Base64.getDecoder().decode(text)

        fun parsePublicKeyPem(pem: String): PublicKey {
            val der = unb64(pem.replace(Regex("-----[A-Z ]+-----|\\s"), ""))
            return KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(der))
        }
    }
}

File: clients/flutter/gateway_client.dart
// clients/flutter/gateway_client.dart
// Dependencies in pubspec.yaml:
//   pointycastle: ^4.0.0
//   basic_utils: ^5.8.2
//   http: ^1.6.0

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:basic_utils/basic_utils.dart';
import 'package:http/http.dart' as http;
import 'package:pointycastle/export.dart';

class GatewayException implements Exception {
  final int status;
  final String? code;
  final String message;
  GatewayException(this.status, this.code, this.message);

  @override
  String toString() => 'GatewayException($status, code=$code): $message';
}

class GatewayResult {
  final int statusCode;
  final Map<String, dynamic> headers;
  final dynamic data;

  GatewayResult({
    required this.statusCode,
    required this.headers,
    required this.data,
  });
}

class GatewayClient {
  final Uri gatewayUrl;
  final String? keyId;
  final bool requireEncryptedResponse;
  final RSAPublicKey _serverKey;
  final http.Client _http;
  final Random _random = Random.secure();

  GatewayClient({
    required this.gatewayUrl,
    required String serverPublicKeyPem,
    this.keyId,
    this.requireEncryptedResponse = true,
    http.Client? httpClient,
  }) : _serverKey = CryptoUtils.rsaPublicKeyFromPem(serverPublicKeyPem),
       _http = httpClient ?? http.Client();

  /// Sends a request through the encryption gateway.
  Future<GatewayResult> send({
    required String method,
    required String accessPoint,
    Map<String, String> headers = const {},
    Object? body,
    Map<String, dynamic>? parameter,
  }) async {
    final payload = {
      'Method': method,
      'AccessPoint': accessPoint,
      'Header': headers.entries
          .map((e) => {'Key': e.key, 'Value': e.value})
          .toList(),
      if (body != null) 'Body': body,
      if (parameter != null) 'Parameter': parameter,
    };

    // 1. Fresh AES-256 key + IV for this request
    final aesKey = _bytes(32);
    final iv = _bytes(12);

    // 2. AES-256-GCM seal the payload (ciphertext + 16-byte tag)
    final sealed = _gcm(true, aesKey, iv, utf8.encode(jsonEncode(payload)));

    // 3. RSA-OAEP-SHA256 encrypt the AES key
    final rsa = OAEPEncoding.withSHA256(RSAEngine())
      ..init(true, PublicKeyParameter<RSAPublicKey>(_serverKey));

    final res = await _http.post(
      gatewayUrl,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        if (keyId != null) 'KeyId': keyId,
        'EncryptedKey': base64Encode(rsa.process(aesKey)),
        'IV': base64Encode(iv),
        'Payload': base64Encode(sealed),
      }),
    );

    final json = _tryJson(res.body);
    if (res.statusCode != 200) {
      final message = json?['message'];
      throw GatewayException(
        res.statusCode,
        json?['code'] as String?,
        message is List ? message.join(', ') : '${message ?? res.body}',
      );
    }

    if (json == null) throw const FormatException('Invalid gateway response');
    final Map<String, dynamic> result;
    if (json['Encrypted'] == false) {
      if (requireEncryptedResponse) {
        throw GatewayException(
          200,
          'ENCRYPTION_REQUIRED',
          'Encrypted response required; check ENCRYPT_RESPONSE on the server',
        );
      }
      if (json.containsKey('IV') || json.containsKey('Payload')) {
        throw const FormatException('Invalid gateway response');
      }
      result = json;
    } else {
      if (json.containsKey('Encrypted') && json['Encrypted'] != true) {
        throw const FormatException('Invalid gateway response');
      }
      // Legacy encrypted envelopes without a marker are also supported.
      final plain = _gcm(
        false,
        aesKey,
        base64Decode(json['IV'] as String),
        base64Decode(json['Payload'] as String),
      );
      result = jsonDecode(utf8.decode(plain)) as Map<String, dynamic>;
    }
    final statusCode = result['StatusCode'];
    if (statusCode is! int ||
        statusCode < 100 ||
        statusCode > 599 ||
        result['Headers'] is! Map<String, dynamic> ||
        !result.containsKey('Data')) {
      throw const FormatException('Invalid gateway response');
    }

    return GatewayResult(
      statusCode: statusCode,
      headers: (result['Headers'] as Map<String, dynamic>?) ?? {},
      data: result['Data'],
    );
  }

  Uint8List _bytes(int n) =>
      Uint8List.fromList(List.generate(n, (_) => _random.nextInt(256)));

  static Uint8List _gcm(
    bool encrypt,
    Uint8List key,
    Uint8List iv,
    List<int> input,
  ) {
    final gcm = GCMBlockCipher(AESEngine())
      ..init(encrypt, AEADParameters(KeyParameter(key), 128, iv, Uint8List(0)));
    return gcm.process(Uint8List.fromList(input));
  }

  static Map<String, dynamic>? _tryJson(String text) {
    try {
      return jsonDecode(text) as Map<String, dynamic>;
    } catch (_) {
      return null;
    }
  }
}
